### ISSUE TITLE
mbsnrtowcs.c: Expression did not evaluate to integer

### ISSUE THREAD

**Issue by brooksdavis** [CTSRD-CHERI/clang#53]
This crash results from trying to switch intptr_t to intcap_t.
https://gist.github.com/brooksdavis/4395ec12e0992501c82c

```
/home/bed22/git/llvm-cheri/Build/bin/clang -integrated-as --target=cheri-unknown-freebsd  -msoft-float --sysroot=/home/bed22/obj/mips.mips64/home/bed22/git/cheribsd/tmp -mabi=sandbox  -O -pipe -G0   -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/include  -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/locale  -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/mips  -I/home/bed22/git/cheribsd/lib/libc_cheri/../../contrib/gdtoa  -I/home/bed22/git/cheribsd/lib/libc_cheri/../../contrib/tzcode/stdtime  -mxgot -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/include                      -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/mips          -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/mips/softfloat        -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/locale               -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/regex                  -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/softfloat                     -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/string                        -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/stdlib                        -I/home/bed22/git/cheribsd/lib/libc_cheri/../libc/stdtime                       -I/home/bed22/git/cheribsd/lib/libc_cheri/../../contrib/gdtoa   -I/home/bed22/git/cheribsd/lib/libc_cheri/../../include -Wno-cast-align -DFORCE_C_LOCALE -DFORCE_UTC_TZ  -std=gnu99 -Wsystem-headers -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wshadow -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wnested-externs -Wredundant-decls -Wold-style-definition -Wno-pointer-sign -Wmissing-variable-declarations -Wthread-safety -Wno-empty-body -Wno-string-plus-int -Wno-unused-const-variable -Qunused-arguments  -c /home/bed22/git/cheribsd/lib/libc_cheri/../libc/locale/mbsnrtowcs.c -o mbsnrtowcs.o
Assertion failed: (EvalResult.Val.isInt() && "Expression did not evaluate to integer"), function EvaluateKnownConstInt, file ../tools/clang/lib/AST/ExprConstant.cpp, line 8600.
clang-3.6: error: unable to execute command: Segmentation fault (core dumped)
clang-3.6: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 3.7.0 (https://github.com/llvm-mirror/clang.git 3da05ea3ad6db2e2d768f7fc72bcf5e6fa35ee14) (https://github.com/llvm-mirror/llvm.git b5ef7b13d0e31e71683126844b9504e267153fac)
Target: cheri-unknown-freebsd
Thread model: posix
clang-3.6: note: diagnostic msg: PLEASE submit a bug report to http://llvm.org/bugs/ and include the crash backtrace, preprocessed source, and associated run script.
clang-3.6: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-3.6: note: diagnostic msg: /tmp/mbsnrtowcs-2ecfc2.c
clang-3.6: note: diagnostic msg: /tmp/mbsnrtowcs-2ecfc2.sh
clang-3.6: note: diagnostic msg: 

********************
*** Error code 254
```

**Comment by davidchisnall** [comment]
This seems to be caused by doing a switch on an `intcap_t`, which then causes problems because the implicit cast of 0 to an `intcap_t` is not a compile-time constant.  This is a bit surprising, because it is a compile-time constant when doing global initialisation.

**Comment by davidchisnall** [comment]
Fixed in 4f61a61
